### ISSUE TITLE
A bug on the command "pdim" ?

### ISSUE THREAD

**Issue by ssario** [Macaulay2/M2#936]
The following statement is taken from the book "A course in ring theory" by D.S.Passman, (page 76).

![statement](https://user-images.githubusercontent.com/49481146/55874490-304e3100-5b9b-11e9-8267-9af599f41cad.png)

I've fixed the bivariate polynomial ring **R** and considered the ideal **I =  < (x^3 + y)^2 , (x^2 + y^2)^2 , (x + y^3)^2 , (x*y)^2 >** . Hence I get the following short exact sequence

**0  >>  I  >>  R  >>  R/I  >>  0**

Here I expect that **pdim (I) = pdim (R/I) - 1** by the third item of the statement above. But when I do the computations on Macaulay2, it gives me **pdim (I) = pdim (R/I) = 2**. Is there a bug on the command pdim? or am I missing something? I've used the following code on M2.

Macaulay2, version 1.13
with packages: ConwayPolynomials, Elimination, IntegralClosure, InverseSystems, LLLBases, PrimaryDecomposition,
               ReesAlgebra, TangentCone, Truncations

i1 : R = QQ[x,y];

i2 : I = ideal((x^3 + y)^2 , (x^2 + y^2)^2 , (x + y^3)^2 , (x*y)^2);

i3 : pdim (module I)

o3 = 2

i4 : pdim (R^1 / I)

o4 = 2

**Comment by mikestillman** [comment]
This is a bug, but one not easily fixed, I think.  The function works correctly for graded modules over graded rings.  By doing ‘code methods dim’, one sees the implementation: it merely returns the length of a free resolution of the module, which in the homogeneous case gives the correct answer (since the returned resolution is minimal in that case).

But for inhomogeneous ideals and modules, we do not compute projective resolutions, only free ones, and ones that are not guaranteed to have minimal length either.

It would be nice to have such a function.  In the meantime, we should change the documentation to note that it doesn’t work for inhomogeneous modules.  Probably, we should either return an error in that case, or at least print a warning that it is only an upper bound on the projective dimension, in the non-graded case.

Best,

Mike

**Comment by eisenbud** [comment]
One can in principle compute the projective dimension from any free resolution by finding the step at which the maximal non-vanishing minors of the differential define the unit ideal. This would be very slow for large examples, but would be fine for the example suggested by Ssario
--
David Eisenbud
Director, Mathematical Sciences Research Institute; and
Professor of Mathematics,University of California, Berkeley
www.msri.org/~de

**Comment by DanGrayson** [comment]
It's better to compute the right answer than to give a warning that the answer might be wrong.  For now, let's just issue a warning message if the module is inhomogeneous.  Who wants to do it?

**Comment by mikestillman** [comment]
The problem is that the methods that @eisenbud suggested will be really slow on reasonable examples.  This might be a good project to add to our (outdated) projects list!

**Comment by DanGrayson** [comment]
I think of this list of issues as a new projects list.

**Comment by mikestillman** [comment]
True.  Maybe we should have a tag ‘project’ for somewhat more involved issues (e.g. this one, where one needs to think through the algorithm).

**Comment by DanGrayson** [comment]
I think they can glean the difficulty from the discussion.

**Comment by DanGrayson** [comment]
Maybe we should just phase out the projects list and shut it down.

**Comment by mikestillman** [comment]
Maybe, but at least one person has mentioned it would be nice to have (sadly, I forget who that was).  They wanted to work on a project, but found that list way out of date.

Keeping the projects as git issues is a good idea though.  Then we can get commentary too, people can comment on if they are working on it, etc.

**Comment by eisenbud** [comment]
What did I suggest??

David
--
David Eisenbud
Director, Mathematical Sciences Research Institute; and
Professor of Mathematics,University of California, Berkeley
www.msri.org/~de

**Comment by kschwede** [comment]
Hi all,

I'm working with an undergraduate (well a couple of them) on a package that should at least improve this.  It will still only give an upper bound, but it should be correct very frequently.  

Essentially we are playing around with grabbing mildly clever submatrices of a matrix and seeing where they vanish (for the purpose of quickly verifying things like R1 or nonsingularity, or birationality, at least in the affirmative).  

One thing we are wondering about is if there are any other good examples where we already know that pdim is not giving us the right answer (length)?

(I guess you can also list me as a volunteer to look at this).  

Karl

**Comment by mikestillman** [comment]
It would be good to address this, @kschwede you mentioned (in 2019!) that maybe `FastMinors` (I guess that was what you were referring to) might help with this?

In any case, the results for `pdim` are wrong for quotient rings (much of the time).  I also see that there is a test in `Pullback.m2` that uses `pdim`, and it gives an error as pdim computes a resolution, without setting a LengthLimit (in the new Complexes code, this is not currently allowed). @kschwede Is it ok if I remove that pdim part of that test?  (It is the only use of pdim in Pullback.m2). Or can we make the test more transparent about what it is about?

edit: this test only fails on the new complexes-26-05 branch, where the new Complexes code replaces the old.  One difference is that free resolutions over quotient rings (and exterior algebras) must be given a LengthLimit.  `pdim` simply uses the default length limit (in the old ChainComplexes case), which is yet another bug in `pdim`!  I would be happy to disallow `pdim` for either inhomogeneous input, or when the ring is a quotient ring.  @eisenbud, @ggsmith @mahrud and Karl (and anyone else listening): what do you think?

**Comment by mahrud** [comment]
My suggestion is to change both pdim and regularity (see #3321) to push forward the module to the ambient polynomial ring first. This is very convenient for geometric applications.

**Comment by mahrud** [comment]
By the way, @mikestillman, see Karl's response in #3356 about projDim in FastMinors. I think adding that as an inhomogeneous strategy would be reasonable. Also see #3656.
